File: 백준/Silver/14425. 문자열 집합/문자열 집합.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());

        TrieNode root = new TrieNode();

        for (int i = 0; i < n; i++) {
            String word = br.readLine();
            insert(root, word);
        }

        int count = 0;
        for (int i = 0; i < m; i++) {
            String word = br.readLine();
            if (search(root, word)) {
                count++;
            }
        }

        System.out.println(count);
    }

    private static void insert(TrieNode root, String word) {
        TrieNode cur = root;

        for (int i = 0; i < word.length(); i++) {
            char c = word.charAt(i);
            int index = c - 'a';

            if (cur.children[index] == null) {
                cur.children[index] = new TrieNode();
            }

            cur = cur.children[index];
        }

        cur.isEndOfWord = true;
    }

    private static boolean search(TrieNode root, String word) {
        TrieNode cur = root;

        for (int i = 0; i < word.length(); i++) {
            char c = word.charAt(i);
            int index = c - 'a';

            if (cur.children[index] == null) {
                return false;
            }

            cur = cur.children[index];
        }

        return cur.isEndOfWord;
    }

    private static class TrieNode {
        TrieNode[] children;
        boolean isEndOfWord;

        public TrieNode() {
            this.children = new TrieNode[26];
            this.isEndOfWord = false;
        }
    }
}

File: 백준/Silver/14425. 문자열 집합/문자열 집합.kt
import java.io.BufferedReader
import java.io.InputStreamReader

fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    val (n, m) = br.readLine().split(" ").map { it.toInt() }

    val root = TrieNode()

    repeat(n) {
        val word = br.readLine()
        insert(root, word)
    }

    var count = 0
    repeat(m) {
        val word = br.readLine()
        if (search(root, word)) {
            count++
        }
    }

    println(count)
}

private fun insert(root: TrieNode, word: String) {
    var cur = root

    for (c in word) {
        val index = c - 'a'

        if (cur.children[index] == null) {
            cur.children[index] = TrieNode()
        }

        cur = cur.children[index]!!
    }

    cur.isEndOfWord = true
}

private fun search(root: TrieNode, word: String): Boolean {
    var cur = root

    for (c in word) {
        val index = c - 'a'

        if (cur.children[index] == null) {
            return false
        }

        cur = cur.children[index]!!
    }

    return cur.isEndOfWord
}

class TrieNode {
    val children = arrayOfNulls<TrieNode>(26)
    var isEndOfWord = false
}
